### PR TITLE
Trie based index for key storage

### PR DESCRIPTION
**Pull request by ya-goodfella**
Trie-based implementation for keys storage.
Provides an average complexity O(|key|) for all operations with worst-case performance proportional to the depth of compressed trie.
Unfortunately, it can be used only for filterable indexes, but it makes lookup operations much faster.
Space requirments remained reasonable comparing to other implementations and depends on dataset:
```
100K random strings with const prefix:
fixed: 2500008 bytes
variable: 3300012 bytes
trie: 2707918 bytes

64K random 8 byte data:
fixed: 524296 bytes
variable: 1048588 bytes
trie: 1961066 bytes

~100K 4 byte integers from range [20000, 250000000):
fixed: 388180 bytes
variable: 1164528 bytes
trie: 1917660 bytes
```

Benchmarks below:
```
Benchmark                                                                                                         Mode   Cnt   Score    Error  Units
ByteArraySortedSetBenchmarks.price_fixed                                                                         thrpt    25  12,137 ±  1,851  ops/s
ByteArraySortedSetBenchmarks.price_trie                                                                          thrpt    25  30,216 ±  1,428  ops/s
ByteArraySortedSetBenchmarks.price_variable                                                                      thrpt    25   8,071 ±  0,871  ops/s
ByteArraySortedSetBenchmarks.random_8bytes_64K_fixed                                                             thrpt    25  18,333 ±  0,523  ops/s
ByteArraySortedSetBenchmarks.random_8bytes_64K_trie                                                              thrpt    25  46,772 ±  2,117  ops/s
ByteArraySortedSetBenchmarks.random_8bytes_64K_variable                                                          thrpt    25  10,669 ±  0,271  ops/s
ByteArraySortedSetBenchmarks.random_prefixed_string_100K_fixed                                                   thrpt    25   5,767 ±  0,175  ops/s
ByteArraySortedSetBenchmarks.random_prefixed_string_100K_trie                                                    thrpt    25  14,527 ±  0,671  ops/s
ByteArraySortedSetBenchmarks.random_prefixed_string_100K_variable                                                thrpt    25   4,304 ±  0,116  ops/s
ByteArraySortedSetBenchmarks.price_fixed                                                                          avgt    25   0,077 ±  0,009   s/op
ByteArraySortedSetBenchmarks.price_trie                                                                           avgt    25   0,035 ±  0,002   s/op
ByteArraySortedSetBenchmarks.price_variable                                                                       avgt    25   0,101 ±  0,002   s/op
ByteArraySortedSetBenchmarks.random_8bytes_64K_fixed                                                              avgt    25   0,054 ±  0,001   s/op
ByteArraySortedSetBenchmarks.random_8bytes_64K_trie                                                               avgt    25   0,023 ±  0,001   s/op
ByteArraySortedSetBenchmarks.random_8bytes_64K_variable                                                           avgt    25   0,096 ±  0,002   s/op
ByteArraySortedSetBenchmarks.random_prefixed_string_100K_fixed                                                    avgt    25   0,180 ±  0,003   s/op
ByteArraySortedSetBenchmarks.random_prefixed_string_100K_trie                                                     avgt    25   0,066 ±  0,002   s/op
ByteArraySortedSetBenchmarks.random_prefixed_string_100K_variable                                                 avgt    25   0,221 ±  0,012   s/op
```

Additionally I fixed bug in LongArrayBitSet.arraySize() and made possible comparision of empty UnsignedByteArray/Buffer.

### CODE REVIEW

File: core/src/main/java/com/yandex/yoctodb/util/common/BufferIterator.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    @Override
+    public Byte next() {
+        if (hasNext()) {
+            return buffer.get(advance++);
+        } else {
+            throw new NoSuchElementException("Empty iterator");
+        }
+    }
+
+    public static int strip(@NotNull Iterator<Byte> from, @NotNull Iterator<Byte> elements) {
```
[the remark this answers is not in the document]
Comment (ya-goodfella) [reply core/src/main/java/com/yandex/yoctodb/util/common/BufferIterator.java]: Итератор - структура одноразовая. В данном случае, метод выполняет функцию compare, но назначение его - отрезать (strip) от from все элементы, которые совпадают с elements, насколько это возможно. Если не получилось - значит from лексикографически либо больше, либо меньше elements - это влияет на выбор веток исполнения дальше.

File: core/src/main/java/com/yandex/yoctodb/util/common/BufferIterator.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
+package com.yandex.yoctodb.util.common;
+
+import com.google.common.primitives.UnsignedBytes;
+import com.yandex.yoctodb.util.buf.Buffer;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+public class BufferIterator implements Iterator<Byte> {
```
[the remark this answers is not in the document]
Comment (ya-goodfella) [reply core/src/main/java/com/yandex/yoctodb/util/common/BufferIterator.java]: Можно попробовать сюда пооптимизировать, спасибо.

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+
+            int next = Byte.toUnsignedInt(query.next());
+            switch (edgeType(metadata)) {
+                case EDGES_SINGLE:
+                    if (next != Byte.toUnsignedInt(nodes.get(movingOffset++))) {
+                        return NOT_FOUND;
+                    }
+                    movingOffset = nodes.getLong(movingOffset);
+                    break;
+                case EDGES_BITSET: {
```
[the remark this answers is not in the document]
Comment (ya-goodfella) [reply core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: Тут (и в кейсе ниже) определены переменные с одинаковым именем. Скобки определяют скоуп имен.

File: core/src/main/java/com/yandex/yoctodb/util/common/BufferIterator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ya-goodfella) [reply core/src/main/java/com/yandex/yoctodb/util/common/BufferIterator.java]: Впрочем, комментарий стоит добавить, спасибо :)

[the remark this answers is not in the document]
Comment (darl) [reply core/src/main/java/com/yandex/yoctodb/util/common/BufferIterator.java]: то что compare калечит итератор, я бы догадался.
а то что strip это на самом деле compare я не сразу понял.

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+     * @param e the element to lookup
+     *
+     * @return index of key matching the query or -1 if there is no match
+     */
+    @Override
+    public int indexOf(@NotNull Buffer e) {
+        if (size() > 0) {
+            return indexOf(new BufferIterator(e));
+        }
+
```
Comment (sanekas) [inline core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: extra \n?

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+     *
+     * @return index of key matching the query or -1 if there is no match
+     */
+    @Override
+    public int indexOfGreaterThan(@NotNull Buffer e, boolean orEquals, int upToIndexInclusive) {
+        if (size() > 0) {
+            int result = indexOfGreaterThan(new BufferIterator(e), orEquals);
+            return result == keysCount ? NOT_FOUND : Math.min(result, upToIndexInclusive);
+        }
+
```
Comment (sanekas) [inline core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: extra \n?

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+                    int max = Byte.toUnsignedInt(nodes.get(movingOffset++));
+
+                    movingOffset += BufferBitSet.arraySize(max - min + 1) * Long.BYTES;
+                    movingOffset = nodes.getLong(movingOffset);
+                    break;
+                case EDGES_CONDENSED:
+                    movingOffset += 2 * Byte.BYTES;
+                    movingOffset = nodes.getLong(movingOffset);
+                    break;
+                default: // should be unreachable
```
Comment (sanekas) [inline core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: but if it is, maybe NOT_FOUND?

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+                    break;
+                }
+                case EDGES_CONDENSED: {
+                    int min = Byte.toUnsignedInt(nodes.get(movingOffset++));
+                    int max = Byte.toUnsignedInt(nodes.get(movingOffset++));
+
+                    movingOffset = nodes.getLong(movingOffset + (max - min) * Long.BYTES);
+                    break;
+                }
+                case EDGES_NONE:
```
Comment (sanekas) [inline core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: default is undefined

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java (hunk 6 above)
Comment (ya-goodfella) [reply core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: There is no correct way to prove this. This case may be unreachable only if a serialized index is corrupted (which may not because we check checksums on the database).

default indicates that I have verified that this case should not happen on regular execution.

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java (hunk 7 above)
Comment (ya-goodfella) [reply core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: It's redundant because `edgeType(int metadata)` is a pure function that may emit only 4 distinct values which have covered by all the cases in this switch.

If you insist, I can change `case EDGES_NONE:` for `default:`, but I don't think that would be any way better for readability.

Comment (ya-goodfella) [reply core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: Anyway, forget about this. Added empty default clause

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java (hunk 4 above)
Comment (ya-goodfella) [reply core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: No, it's just like it meant to be.

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java (hunk 5 above)
Comment (ya-goodfella) [reply core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: No, it's just like it meant to be.

File: core/src/main/java/com/yandex/yoctodb/util/common/BufferIterator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ya-goodfella) [reply core/src/main/java/com/yandex/yoctodb/util/common/BufferIterator.java]: Переделал интерфейс, спасибо. Посмотри, стало ли понятнее?

File: core/src/main/java/com/yandex/yoctodb/util/common/BufferIterator.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ya-goodfella) [reply core/src/main/java/com/yandex/yoctodb/util/common/BufferIterator.java]: Сильного выигрыша не дало - замеры в рамках погрешности; но выглядит так почище.

File: core/src/main/java/com/yandex/yoctodb/mutable/DocumentBuilder.java [hunk 8]
```diff
@@ -32,7 +32,8 @@
 
     enum LengthOption {
         FIXED,
-        VARIABLE
+        VARIABLE,
+        TRIE
```
Comment (incubos) [inline core/src/main/java/com/yandex/yoctodb/mutable/DocumentBuilder.java]: It is quite unexpected to see this `enum` being extended...

File: core/src/main/java/com/yandex/yoctodb/util/common/BufferIterator.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+            int result = Integer.compare(this.next(), prefix.next());
+            if (result != 0) {
+                return result;
+            }
+        }
+
+        return prefix.hasNext() ? -1 : 0;
+    }
+
+    public static BufferIterator wrapCopy(@NotNull final Collection<Byte> bytes) {
```
Comment (incubos) [inline core/src/main/java/com/yandex/yoctodb/util/common/BufferIterator.java]: Is it possible to switch to `byte[]` instead of much less efficient `Collection<Byte>` everywhere (if it is an issue)?

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/BufferBitSet.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+            if (++u == wordCount) {
+                return -1;
+            }
+            word = buf.getLong(bufferOffset + u * Long.BYTES);
+        }
+    }
+
+    public static int cardinalityTo(@NotNull final Buffer buf,
+                                    final long bufferOffset,
+                                    final int i) {
```
Comment (incubos) [inline core/src/main/java/com/yandex/yoctodb/util/immutable/impl/BufferBitSet.java]: I'd rename `i` into something more meaningful.

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (incubos) [reply core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: Для единообразия стоит во всех case'ах фигурные скобки использовать.

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java [hunk 11]
```diff
@@ -0,0 +1,444 @@
...
+                movingOffset += Integer.BYTES;
+            }
+
+            if (!query.hasNext()) {
+                return maybeValue;
+            }
+
+            final int next = query.next();
+            switch (edgeType(metadata)) {
+                case EDGES_SINGLE:
```
Comment (incubos) [inline core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: It's worth documenting these `enum`s in the definition.

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java [hunk 12]
```diff
@@ -0,0 +1,444 @@
...
+                    int max = Byte.toUnsignedInt(nodes.get(movingOffset++));
+                    if (min > next || max < next) {
+                        return NOT_FOUND;
+                    }
+                    int index = next - min;
+                    movingOffset = nodes.getLong(movingOffset + index * Long.BYTES);
+                    break;
+                }
+                default:
+                    return NOT_FOUND;
```
Comment (incubos) [inline core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: Isn't that an error (`UnsupportedOperationException`)?

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java [hunk 13]
```diff
@@ -0,0 +1,444 @@
...
+    @Override
+    public int indexOfLessThan(@NotNull Buffer e, boolean orEquals, int fromIndexInclusive) {
+        if (size() > 0) {
+            return indexOfLessThan(new BufferIterator(e), orEquals);
+        }
+
+        return NOT_FOUND;
+    }
+
+    private int indexOf(@NotNull final BufferIterator query) {
```
Comment (incubos) [inline core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: I can't grasp what is going on here, actually :/

Comment (incubos) [reply core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: Please, don't shy away from injecting numerous comments :)

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java [hunk 14]
```diff
@@ -0,0 +1,444 @@
...
+        }
+    }
+
+    private int takeFirstValueOnLeft(long movingOffset) {
+        while (true) {
+            final int metadata = Byte.toUnsignedInt(nodes.get(movingOffset));
+            movingOffset += Byte.BYTES;
+
+            if (hasPrefix(metadata)) { // there is an prefix
+                int prefixSize = nodes.getInt(movingOffset);
```
Comment (incubos) [inline core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: `final`?

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java (hunk 6 above)
Comment (incubos) [reply core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: I'd throw `UnsupportedOperationException` if it is possible to maintain 100% coverage.

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java (hunk 7 above)
Comment (incubos) [reply core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: Empty `default` case is quite a risk when maintaining the code.

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java [hunk 15]
```diff
@@ -0,0 +1,444 @@
...
+                    int key = Byte.toUnsignedInt(nodes.get(movingOffset++));
+                    if (next > key) {
+                        return takeLastValueOnRight(nodes.getLong(movingOffset));
+                    } else if (next < key) {
+                        return takeFirstValueOnLeft(nodes.getLong(movingOffset)) - 1;
+                    }
+                    movingOffset = nodes.getLong(movingOffset);
+                    break;
+                case EDGES_BITSET: {
+                    int min = Byte.toUnsignedInt(nodes.get(movingOffset++));
```
Comment (incubos) [inline core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: I am not quite sure, but is it possible to refactor this extremely similar code into separate `private` methods?

File: core/src/main/java/com/yandex/yoctodb/util/mutable/impl/LongArrayBitSet.java [hunk 16]
```diff
@@ -46,7 +46,7 @@ private LongArrayBitSet(
     }
 
     static int arraySize(final int size) {
-        return (size >>> 6) + 1;
+        return (size >>> 6) + ((size & 0x3f) != 0 ? 1 : 0);
```
Comment (incubos) [inline core/src/main/java/com/yandex/yoctodb/util/mutable/impl/LongArrayBitSet.java]: :+1:

File: core/src/main/java/com/yandex/yoctodb/util/mutable/impl/LongArrayBitSet.java [hunk 17]
```diff
@@ -77,6 +77,8 @@ public static ArrayBitSet one(final int size) {
         final int shift = size & 0x3f;
         if (shift != 0) {
```
Comment (incubos) [inline core/src/main/java/com/yandex/yoctodb/util/mutable/impl/LongArrayBitSet.java]: It might be better to replace branches and switch to `==`.

File: core/src/main/java/com/yandex/yoctodb/util/mutable/impl/TrieByteArraySortedSet.java [hunk 18]
```diff
@@ -0,0 +1,279 @@
...
+        private Integer value = null;
+        private long offset;
+
+        void append(UnsignedByteArray key, int value) {
+            checkNotFrozen();
+            append(key.iterator(), value);
+        }
+
+        private void compress() {
+            while (edges.size() == 1 && value == null) {
```
Comment (incubos) [inline core/src/main/java/com/yandex/yoctodb/util/mutable/impl/TrieByteArraySortedSet.java]: Extensive JavaDocs/comments are welcome everywhere.

File: core/src/main/java/com/yandex/yoctodb/util/mutable/impl/TrieByteArraySortedSet.java [hunk 19]
```diff
@@ -0,0 +1,279 @@
...
+        private int getMax() {
+            return edges.lastKey();
+        }
+
+        private int bitSetSizeInLongs() {
+            return LongArrayBitSet.arraySize(getMax() - getMin() + 1);
+        }
+
+        private boolean isCondensed() {
+            return !edges.isEmpty() && getMax() - getMin() == edges.size() - 1;
```
Comment (incubos) [inline core/src/main/java/com/yandex/yoctodb/util/mutable/impl/TrieByteArraySortedSet.java]: I'd add brackets around separate expressions to make it more readable.

File: core/src/main/java/com/yandex/yoctodb/util/mutable/impl/TrieByteArraySortedSet.java [hunk 20]
```diff
@@ -0,0 +1,279 @@
...
+        }
+
+        private void writeNodeTo(@NotNull OutputStream os) throws IOException {
+            writeMetadata(os);
+
+            if (prefix.size() > 0) {
+                os.write(Ints.toByteArray(prefix.size())); // prefix size
+                for (byte b : prefix) {
+                    os.write(b);
+                } // prefix bytes
```
Comment (incubos) [inline core/src/main/java/com/yandex/yoctodb/util/mutable/impl/TrieByteArraySortedSet.java]: Move the comment above `for`?

File: core/src/main/java/com/yandex/yoctodb/util/mutable/impl/TrieByteArraySortedSet.java [hunk 21]
```diff
@@ -0,0 +1,279 @@
...
+            }
+
+            if (value != null) {
+                os.write(Ints.toByteArray(value)); // value
+            }
+
+            if (edges.size() == 1) {
```
Comment (incubos) [inline core/src/main/java/com/yandex/yoctodb/util/mutable/impl/TrieByteArraySortedSet.java]: These alternatives should be described in JavaDoc somewhere.

File: core/src/main/java/com/yandex/yoctodb/v1/V1DatabaseFormat.java [hunk 22]
```diff
@@ -29,8 +29,6 @@
  */
 @ThreadSafe
 public final class V1DatabaseFormat extends DatabaseFormat {
-    public final static int FORMAT = 7;
```
Comment (incubos) [inline core/src/main/java/com/yandex/yoctodb/v1/V1DatabaseFormat.java]: Why is that removed?

File: core/src/test/java/com/yandex/yoctodb/util/common/BufferIteratorTest.java [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+
+import java.nio.ByteBuffer;
+import java.util.NoSuchElementException;
+
+public class BufferIteratorTest {
+    @Test(expected = NoSuchElementException.class)
+    public void invalidNext() {
+        new BufferIterator(Buffer.from(ByteBuffer.allocate(0))).next();
+    }
+}
```
Comment (incubos) [inline core/src/test/java/com/yandex/yoctodb/util/common/BufferIteratorTest.java]: `\n` missing.

File: core/src/test/java/com/yandex/yoctodb/util/immutable/impl/BufferBitSetTest.java [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+        assertEquals(1, BufferBitSet.cardinalityTo(bitsetBuffer, 0, 1));
+        assertEquals(32, BufferBitSet.cardinalityTo(bitsetBuffer, 0, 63));
+        assertEquals(31, BufferBitSet.cardinalityTo(bitsetBuffer, 0, 62));
+        assertEquals(33, BufferBitSet.cardinalityTo(bitsetBuffer, 0, 66));
+    }
+
+    @Test
+    public void arraySize() {
+    }
+}
```
Comment (incubos) [inline core/src/test/java/com/yandex/yoctodb/util/immutable/impl/BufferBitSetTest.java]: `\n` missing.

File: core/src/test/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSetTest.java [hunk 25]
```diff
@@ -0,0 +1,263 @@
...
+        trieSet = com.yandex.yoctodb.util.immutable.impl.TrieByteArraySortedSet.from(
+                Buffer.from(trieOutput.toByteArray())
+        );
+        variableSet = com.yandex.yoctodb.util.immutable.impl.VariableLengthByteArraySortedSet.from(
+                Buffer.from(variableOutput.toByteArray())
+        );
+
+        queries.addAll(falseQueries);
+    }
+
```
Comment (incubos) [inline core/src/test/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSetTest.java]: Extra empty lines.

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java (hunk 7 above)
Comment (ya-goodfella) [reply core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: That's interesting. I thought that an empty default case is meaningless for a program flow and Java would treat it as if it's not here, and it's only indicating for the code reader that I have thought about that option and leaved it with exactly that implementation. Could you please extend your opinion on that?

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java (hunk 15 above)
Comment (ya-goodfella) [reply core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: I think it would be hard without losing performance - that line generates value and also moves an external buffer pointer.

That code is written in while loop with moving pointer to achieve stack safety - that's some tail recursion emulation :)

File: core/src/main/java/com/yandex/yoctodb/v1/V1DatabaseFormat.java (hunk 22 above)
Comment (ya-goodfella) [reply core/src/main/java/com/yandex/yoctodb/v1/V1DatabaseFormat.java]: Because we moved to features description in previous pull request - but forgotten to good ol' FORMAT that left unused.

File: core/src/main/java/com/yandex/yoctodb/mutable/DocumentBuilder.java (hunk 8 above)
Comment (ya-goodfella) [reply core/src/main/java/com/yandex/yoctodb/mutable/DocumentBuilder.java]: Yep, it feels not very natural. Looks like we need to introduce another way to control index serialization on a time index being created, what do you think?

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java (hunk 7 above)
Comment (incubos) [reply core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: Just imaging extending the `enum` when somebody adds a feature in future. `default` empty branch will silently swallow the case which may lead to unexpected results.

File: core/src/main/java/com/yandex/yoctodb/mutable/DocumentBuilder.java (hunk 8 above)
Comment (incubos) [reply core/src/main/java/com/yandex/yoctodb/mutable/DocumentBuilder.java]: Let's rename the `enum` to `IndexType` and `FIXED`/`VARIABLE` to something more appropriate.

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java (hunk 15 above)
Comment (incubos) [reply core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java]: Holy, shit...

File: core/src/main/java/com/yandex/yoctodb/util/common/TrieNodeMetadata.java [hunk 26]
```diff
@@ -10,16 +10,19 @@
 
 package com.yandex.yoctodb.util.common;
 
+/**
+ *
```
Comment (incubos) [inline core/src/main/java/com/yandex/yoctodb/util/common/TrieNodeMetadata.java]: ?

File: core/src/main/java/com/yandex/yoctodb/util/immutable/impl/BufferBitSet.java [hunk 27]
```diff
@@ -58,20 +58,27 @@ public static int nextSetBit(@NotNull final Buffer buf,
         }
     }
 
+    /**
+     * Counts bits from 0 until {@code bitIndex}
```
Comment (incubos) [inline core/src/main/java/com/yandex/yoctodb/util/immutable/impl/BufferBitSet.java]: You can move this line to `@return` or remove `@return` altogether.
